### ISSUE TITLE
Runtime dependencies when using a statically linked driver

### ISSUE THREAD

**Issue by diml** [whitequark/ppx_deriving#96]
I'm currently refreshing the doc for ppx\_driver and adding tests for other jane street rewriters. At the moment there is one annoying issue when using a statically linked ppx driver and I believe the same issue exist when using a static ppx\_deriving driver.  This ticket is to discuss a common solution that could be advertised by both libraries.

The issue is that files that are pre-processed using a custom driver still needs the runtime dependencies of the various rewriters. Sometimes these are obvious and it's likely users will manually write them in their build system configuration (such as `sexplib`, `yojson`, ...) so it's not a big deal. However sometimes the runtime dependencies are some internal stuff that the user doesn't necessarily knows about.

For instance let's say that you are using a rewriter `ppx_foo` statically linked in `./custom-ppx` and `ppx_foo` has some obscure runtime dependencies;
- If you pass only `-ppx ./custom-ppx` to the compilation commands, then you don't get the runtime dependencies, which is a problem
- if you pass both `-package ppx_foo` and `-ppx ./custom-ppx` you get the runtime dependencies, but it's likely that `-package ppx_foo` will expand into a `-ppx <some-path>/ppx_foo/ppx`, which is unfortunate at best (running the same rewriter twice) and can even be bad (code generated twice)

For convenience though, it's nicer if user can just keep on passing `-package ppx_foo` and get the right behavior. I have two solutions in mind:

#### 1. Always define a `.runtime` sub-package

This way whenever you are using a custom driver, you just replace all ppx packages by `ppx_package.runtime` without thinking about it in your build configuration. Sometimes these packages will be empty to be _future proof_.

This method is a bit tedious as if you miss adding the `.runtime` for one ppx package, you might get into a state that is hard to debug.

#### 2. Add a `use_ppx_driver` predicate

The idea is that ppx rewriters authors must modify the `ppxopt` variable in their META as follow:

```
ppxopt(-ppx_driver) = blah
-->
ppxopt(-ppx_driver,-use_ppx_driver) = blah
```

And users just put `<src/*>: predicate(use_ppx_driver)` in their `_tags` file.

I think (2) is a better solution as it's easier to switch between the two alternatives, and it would play better with plugins like [ppx_driver.ocamlbuild](https://github.com/janestreet/ppx_driver/blob/master/ocamlbuild_plugin/ppx_driver_ocamlbuild.ml) that build such custom drivers automatically.

Thoughts?

**Comment by whitequark** [comment]
I'm in favor of option 2, but I feel that the `use_ppx_driver` name is rather confusing. It took me a while to untangle what exactly it means. I think something to the extent of "we run the AST rewriters ourselves" would be more meaningful--`external_ppx_rewriting`? `no_integrated_ppx_rewriting`? (I don't like these two either...)

**Comment by diml** [comment]
Maybe `custom_ppx`?

**Comment by whitequark** [comment]
Sounds reasonably good.

**Comment by diml** [comment]
Alright, I'm using this tag for the upcoming minor release

**Comment by whitequark** [comment]
Reopening since ppx_deriving (also yojson and protobuf) do not yet follow the convention.

**Comment by whitequark** [comment]
Fixed in e9cbb151be.

@diml, can you please take a look and confirm that it works as you expect?

**Comment by diml** [comment]
I think it's all good, thanks
